From 8b83805103aa0d4e57f2ff8ffe12656a77087f15 Mon Sep 17 00:00:00 2001
From: Xaver Hugl <xaver.hugl@kde.org>
Date: Mon, 17 Aug 2026 16:19:32 +0200
Subject: [PATCH] backends/drm: filter out unusably small modes

If a user accidentally selects a mode like 320x240, they will have a
very hard time using system settings to get back to a usable mode.
To avoid that, this simply filters out all modes smaller than 480p,
unless all modes are that small.
---
 src/backends/drm/drm_connector.cpp | 15 +++++++++++++--
 1 file changed, 13 insertions(+), 2 deletions(-)

diff --git a/src/backends/drm/drm_connector.cpp b/src/backends/drm/drm_connector.cpp
index 55a5f1f8c0..e857de9e5b 100644
--- a/src/backends/drm/drm_connector.cpp
+++ b/src/backends/drm/drm_connector.cpp
@@ -313,8 +313,19 @@ bool DrmConnector::updateProperties()
     if (!equal && m_conn->count_modes > 0) {
         // reload modes
         m_driverModes.clear();
-        for (int i = 0; i < m_conn->count_modes; i++) {
-            m_driverModes.append(std::make_shared<DrmConnectorMode>(this, m_conn->modes[i], OutputModeline::Flags()));
+        const auto modes = std::span(m_conn->modes, m_conn->count_modes);
+        auto usableModes = modes | std::views::filter([](drmModeModeInfo mode) {
+            return mode.hdisplay >= 640 && mode.vdisplay >= 480;
+        });
+        if (usableModes.empty()) {
+            // allow unusable modes, we don't have much of a choice
+            for (const auto &mode : modes) {
+                m_driverModes.append(std::make_shared<DrmConnectorMode>(this, mode, OutputModeline::Flags()));
+            }
+        } else {
+            for (const auto &mode : usableModes) {
+                m_driverModes.append(std::make_shared<DrmConnectorMode>(this, mode, OutputModeline::Flags()));
+            }
         }
         m_modes.clear();
         m_modes.append(m_driverModes);
-- 
2.55.0

